File: Cpp/0140_word_break_ii.cpp
// https://leetcode.cn/problems/word-break/

class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        string sentence;
        backtrack(s, wordDict, sentence, 0);
        return res;
    }

private:
    void backtrack(string s, vector<string>& wordDict, string sentence, int idx) {
        if (idx == s.size()) {
            sentence.pop_back();
            res.push_back(sentence);
            return;
        }

        for (int length = 1; length <= s.size()-idx; ++length) {
            string sub = s.substr(idx, length);
            if (find(wordDict.begin(), wordDict.end(), sub) != wordDict.end()) {
                backtrack(s, wordDict, sentence+sub+" ", idx+length);
            }
        }
    }

    vector<string> res;
};

File: Python3/0140_word_break_ii.py
# https://leetcode.cn/problems/word-break-ii/

class Solution:
    def wordBreak(self, s, wordDict):
        """
        Inputs:
            s: str
            wordDict: list[str]
        Outputs:
            res: list[str]
        """

        self.res = []
        self.backtrack(s, wordDict, [], 0)
        return self.res

    def backtrack(self, s, wordDict, sentence, idx):
        if idx == len(s):
            self.res.append(" ".join(sentence))
            return

        for length in range(1, len(s)+1-idx):
            sub = s[idx: idx+length]
            if sub in wordDict:
                self.backtrack(s, wordDict, sentence+[sub], idx+length)
